### PR TITLE
Mariacarlina/new library version

### PR DESCRIPTION
**Pull request by mariacarlinahernandez**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -19,106 +19,73 @@ These modules use TTL Serial to communicate, so at least 2 pins are required to
...
 
 ## Run the examples!
 
 1. Download the UbidotsFONA library [here](https://github.com/ubidots/Ubidots-FONA/archive/1.2.0.zip)
 2. Go to the Arduino IDE, click on Sketch -> Include Library -> Add .ZIP Library
 3. Select the UbidotsFONA .ZIP file and then "Accept" or "Choose"
-4. Do the same with the Adafruit_FONA library.
-5. Close the Arduino IDE and open it again.
+4. Do the same with the [Adafruit_FONA](https://github.com/adafruit/Adafruit_FONA) library.
+5. Close the Arduino IDE and open it again.Pin Connections FONA -> Arduino
```
Comment (jotathebest) [inline README.md]: Put a blank space after "again".

File: README.md [hunk 2]
```diff
@@ -19,106 +19,73 @@ These modules use TTL Serial to communicate, so at least 2 pins are required to
...
+
 **Before continuing, make sure your device can connect to the Internet! You can run Adafruit's example "FONAtest.ino" contained in [Adafruit's FONA](https://github.com/adafruit/Adafruit_FONA_Library/archive/1.3.0.zip) library to make sure your FONA is working properly and is able to connect to your mobile network.**
 7. Now go to **File -> Examples -> UbidotsFONA** library and select one example.
-6. Put your Ubidots token and variable ID
+6. Put your Ubidots token, device label and variable label where is indicated.
```
Comment (jotathebest) [inline README.md]: "where indicated"

File: README.md [hunk 3]
```diff
@@ -19,106 +19,73 @@ These modules use TTL Serial to communicate, so at least 2 pins are required to
...
 7. Put your mobile operator's APN settings (APN, USER, PASSWORD). You should be able to easily find your operator's APN settings on Google or in the operator's website.
```
Comment (jotathebest) [inline README.md]: Put your mobile operator's APN credentials (APN, USER, PASSWORD). You should be able to easily find your operator's APN settings on Google or at the operator's website.

File: README.md [hunk 4]
```diff
@@ -19,106 +19,73 @@ These modules use TTL Serial to communicate, so at least 2 pins are required to
...
+### Ubidots
+`setDeviceName(char *device_name)`
+> Assign a new device name. The device name by default is **"FONA"**
 
-#define APN  ""  // The APN of your operator
-#define USER ""  // if your apn doesnt have username just leave it ""
-#define PASS ""  // if your apn doesnt have password just leave it ""
-#define TOKEN "Your_Token_Here"  // Replace it with your Ubidots token
+`void setDeviceLabel(char *device_label)`
+> Assign a new device label. The device label by default is **"FONA"**
```
Comment (jotathebest) [inline README.md]: Add here which characters are not allowed inside the device label

File: README.md [hunk 5]
```diff
@@ -19,106 +19,73 @@ These modules use TTL Serial to communicate, so at least 2 pins are required to
...
 
-Ubidots client(TOKEN);
+`add(char *variable_label, double value)`
+> Add a variable with a value
```
Comment (jotathebest) [inline README.md]: Make this description more clear

File: README.md [hunk 6]
```diff
@@ -19,106 +19,73 @@ These modules use TTL Serial to communicate, so at least 2 pins are required to
...
-    sprintf(context_2, "Color=RED$Position=Down$Quantity=3");
-    // You can send multiple context in one variable, to send it you must add a "$" symbol between every context
-    client.add("Temperature", value1, context);  // Change for your variable name
-    client.add("Humedity", value2,  context_2);
-    client.sendAll();
-    delay(5000);
-}
-```
+`add(char *variable_label, double value, char *ctext)`
+> Add a variable with a value and context [optional]
```
Comment (jotathebest) [inline README.md]: Add here an example of a valid context

File: README.md [hunk 7]
```diff
@@ -19,106 +19,73 @@ These modules use TTL Serial to communicate, so at least 2 pins are required to
...
 
+`add(char *variable_label, double value, unsigned long timestamp_val)`
+> Add a variable with a value and timestamp [optional]
```
Comment (jotathebest) [inline README.md]: timestamp should be in seconds or milliseconds?

File: README.md [hunk 8]
```diff
@@ -19,106 +19,73 @@ These modules use TTL Serial to communicate, so at least 2 pins are required to
...
 
-## Get a value from a Ubidots variable
+`add(char *variable_label, double value, char *ctext, unsigned long timestamp)`
+> Add a variable with a value, context, and timestamp [optional]
```
Comment (jotathebest) [inline README.md]: add here a valid example

File: README.md [hunk 9]
```diff
@@ -19,106 +19,73 @@ These modules use TTL Serial to communicate, so at least 2 pins are required to
...
 
-To get a value you could use the example inside the library or copy the following code:
+`sendAll()`
+> Send all data of all variables that you saved using the method **add(char *variable_label, double value)**
```
Comment (jotathebest) [inline README.md]: Does this method returns something or not? Please clarify this

File: README.md [hunk 10]
```diff
@@ -19,106 +19,73 @@ These modules use TTL Serial to communicate, so at least 2 pins are required to
...
 
-```c
-#include <UbidotsFONA.h>
-#include <SoftwareSerial.h>
+`getValue(char* device_label, char* variable_label)`
+> Return the last value from a specific variable
```
Comment (jotathebest) [inline README.md]: what type does this return? float, int, double?

File: src/UbidotsFONA.cpp [hunk 11]
```diff
@@ -0,0 +1,521 @@
...
+    _currentValue = 0;
+    val = (Value *)malloc(MAX_VALUES*sizeof(Value));
+}
+
+/***************************************************************************
+FONA MODULE FUNCTIONS
+***************************************************************************/
+
+bool Ubidots::begin() {
+    pinMode(FONA_RST, OUTPUT);
```
Comment (jotathebest) [inline src/UbidotsFONA.cpp]: Are these delays necessary? If yes, print some debugs messages before every delay

File: src/UbidotsFONA.cpp [hunk 12]
```diff
@@ -0,0 +1,521 @@
...
+    digitalWrite(FONA_RST, HIGH);
+    delay(500);
+    digitalWrite(FONA_RST, LOW);
+    delay(500);
+    digitalWrite(FONA_RST, HIGH);
+    delay(500);
+    Serial.println(F("Attempting to open comm with ATs"));
+    int16_t timeout = 10000;
+    while (timeout > 0) {
+        if (sendMessageAndwaitForOK("AT"))
```
Comment (jotathebest) [inline src/UbidotsFONA.cpp]: put braces here

File: src/UbidotsFONA.cpp [hunk 13]
```diff
@@ -0,0 +1,521 @@
...
+        Serial.println(F("Timeout: No response to AT... last ditch attempt."));
+        sendMessageAndwaitForOK("AT");
+        delay(100);
+        sendMessageAndwaitForOK("AT");
+        delay(100);
+        sendMessageAndwaitForOK("AT");
+        delay(100);
+        return false;
+    }
+    sendMessageAndwaitForOK("ATE0");
```
Comment (jotathebest) [inline src/UbidotsFONA.cpp]: what happens if this returns false?

File: src/UbidotsFONA.cpp [hunk 14]
```diff
@@ -0,0 +1,521 @@
...
+/**
+ * Send all data of all variables that you saved
+ * @reutrn true upon success, false upon error.
+ */
+bool Ubidots::sendAll() {
+    int i;
+    char* allData = (char *) malloc(sizeof(char) * 700);
+    char str_values[10];
+
+    if (_device_name != NULL) {
```
Comment (jotathebest) [inline src/UbidotsFONA.cpp]: Is it possible that _device_name can be NULL? It is initialized as "FONA"

File: src/UbidotsFONA.cpp [hunk 15]
```diff
@@ -0,0 +1,521 @@
...
+      sprintf(allData, "%s/%s|POST|%s|%s:%s=>", USER_AGENT, VERSION, _token, _device_label, _device_name);
+    } else {
+      sprintf(allData, "%s/%s|POST|%s|%s=>", USER_AGENT, VERSION, _token, _device_label);
+    }
+
+    for (i = 0; i < _currentValue; ) {
+        dtostrf(((val+i)->varValue), 4, 4, str_values);
```
Comment (jotathebest) [inline src/UbidotsFONA.cpp]: set packet size to 6 and accuracy to 2

File: src/UbidotsFONA.cpp [hunk 16]
```diff
@@ -0,0 +1,521 @@
...
+        _currentValue = MAX_VALUES;
+  }
+}
+
+/**
+ * Send all data of all variables that you saved
+ * @reutrn true upon success, false upon error.
+ */
+bool Ubidots::sendAll() {
+    int i;
```
Comment (jotathebest) [inline src/UbidotsFONA.cpp]: declare this a uint8_t

File: src/UbidotsFONA.cpp [hunk 17]
```diff
@@ -0,0 +1,521 @@
...
+        Serial.println(F("Error with AT+CIPSEND"));
+#endif
+        return false;
+    }
+    fonaSS.println(allData);
+    char* resp = readData(4000);
+    response = strtok(resp, "|");
+    serverResponse = response;
+    while (response!=NULL) {
+        printf("%s\n", response);
```
Comment (jotathebest) [inline src/UbidotsFONA.cpp]: this should be inside this guard: #ifdef DEBUG_UBIDOTS

File: src/UbidotsFONA.cpp [hunk 18]
```diff
@@ -0,0 +1,521 @@
...
+#ifdef DEBUG_UBIDOTS
+        Serial.println(F("Error with AT+CIPSHUT"));
+#endif
+        return false;
+    }
+    free(allData);
+    return num;
+}
+
+int Ubidots::readLine(uint32_t ts_max, bool multiline) {
```
Comment (jotathebest) [inline src/UbidotsFONA.cpp]: Where are you using this method? I see it declared but it is not being used anywhere

File: src/UbidotsFONA.cpp [hunk 19]
```diff
@@ -0,0 +1,521 @@
...
+      dataLen++;
+    } else {
+      break;
+    }
+  }
+  return dataLen;
+}
+
+
+// ------------------------------------------------------------------
```
Comment (jotathebest) [inline src/UbidotsFONA.cpp]: If these functions were deprecated, do not add them. We grow up the number version to 3.0 so we should not include any deprecated function

File: src/UbidotsFONA.h [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+    bool setApn(char* apn, char* user = "", char* pwd = "");
+    bool sendAll();
+    void add(char *variable_label, double value);
+    void add(char *variable_label, double value, char *ctext);
+    void add(char *variable_label, double value, unsigned long timestamp_val);
+    void add(char *variable_label, double value, char *ctext, unsigned long timestamp);
+    void setDeviceName(char *device_name);
+    void setDeviceLabel(char *device_label);
+    float getValue(char* device_label, char* variable_label);
+    // Deprecated functions
```
Comment (jotathebest) [inline src/UbidotsFONA.h]: do not add deprecated functions

File: src/UbidotsFONA.h [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+    bool isTimedOut(uint32_t ts) { return (long)(millis() - ts) >= 0; }
+    bool begin();
+    char* readData(uint16_t timeout);
+    char* _token;
+    char* _dsName;
+    char* _dsTag;
+    char* _device_label;
+    char* _device_name;
+    char* _server;
+    char buffer[DEFAULT_BUFFER_SIZE];
```
Comment (jotathebest) [inline src/UbidotsFONA.h]: this variables is being used at readLine() but you do not call this method anywhere, erase it

File: src/UbidotsFONA.h [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+
+#ifndef __UbidotsFONA_H_
+#define __UbidotsFONA_H_
+#define DEBUG_UBIDOTS
+
+#include <SoftwareSerial.h>
+#include <stdlib.h>
+#include "Arduino.h"
+
+#define DEFAULT_BUFFER_SIZE      64
```
Comment (jotathebest) [inline src/UbidotsFONA.h]: are you  using this?

File: examples/UbidotsGetValue/UbidotsGetValue.ino [hunk 23]
```diff
@@ -1,24 +1,24 @@
 #include <UbidotsFONA.h>
 #include <SoftwareSerial.h>
 
-
```
Comment (jotathebest) [inline examples/UbidotsGetValue/UbidotsGetValue.ino:4]: Add in all the examples how to activate debug messages

File: examples/UbidotsGetValue/UbidotsGetValue.ino [hunk 24]
```diff
@@ -1,24 +1,24 @@
...
   Serial.begin(115200);
   delay(2000);
   while(!client.setApn(APN, USER, PASS));
 }
 
 void loop() {
-  float value = client.getValueWithDatasource("Data_Source_Label", "Variable_Label");
-  delay(1000); 
+  float value = client.getValue(DEVICE_LABEL, VARIABLE_LABEL);
+  delay(1000);
```
Comment (jotathebest) [inline examples/UbidotsGetValue/UbidotsGetValue.ino]: Is this delay necessary?  There are a lot of 4000 seconds delay inside the getValue() method

File: examples/UbidotsSendMultipleValues/UbidotsSendMultipleValues.ino [hunk 25]
```diff
@@ -2,25 +2,30 @@
...
   float value1 = analogRead(A0);
   float value2 = analogRead(A1);
   float value3 = analogRead(A2);
-  client.add("Variable_Label_1", value1);
-  client.add("Variable_Label_2", value2);
-  client.add("Variable_Label_3", value3);
+  client.add(VARIABLE_LABEL_1, value1);
+  client.add(VARIABLE_LABEL_2, value2);
+  client.add(VARIABLE_LABEL_3, value3);
   client.sendAll();
```
Comment (jotathebest) [inline examples/UbidotsSendMultipleValues/UbidotsSendMultipleValues.ino]: if(client.sendAll()){
    Serial.println("values sent properly");
}

File: examples/UbidotsSendValuesWithContext/UbidotsSendValuesWithContext.ino [hunk 26]
```diff
@@ -3,30 +3,28 @@
...
+    char lng[] = "-122.431297";
     // To send latitude and longitude to Ubidots and see it in a map widget
-    // The format of the context is like this, you must send it like this example
-    sprintf(context_2, "Temperature=Value_Of_Temperature$Position=Value_Of_Position$Color=Value_Of_Color");
-    // You can send multiple context in one variable, to send it you must add a "$" symbol between every context
-    client.add("Variable_Name_One", value1, context);  // Change for your variable name
-    client.add("Variable_Name_Two", value2,  context_2);
+    sprintf(context, "lat=%s$lng=%s", lat, lng);
+    client.add(VARIABLE_LABEL, 1.00 , context);
     client.sendAll();
```
Comment (jotathebest) [inline examples/UbidotsSendValuesWithContext/UbidotsSendValuesWithContext.ino]: if(client.sendAll()){
Serial.println("values sent properly");
}

File: examples/UbidotsSendValuesWithContext/UbidotsSendValuesWithContext.ino [hunk 27]
```diff
@@ -3,30 +3,28 @@
...
     delay(5000);
```
Comment (jotathebest) [inline examples/UbidotsSendValuesWithContext/UbidotsSendValuesWithContext.ino]: Is this delay necessary? there are a lot of delays inside the sendAll() method

File: examples/UbidotsGetValue/UbidotsGetValue.ino (hunk 23 above)
Comment (mariacarlinahernandez) [reply examples/UbidotsGetValue/UbidotsGetValue.ino:4]: The debug method was included.

**Pull request merged**